kubernetes/client_factory.go

```diff
@@ -0,0 +1,132 @@
+package kubernetes
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/kiali/kiali/log"
+	"github.com/kiali/kiali/prometheus/internalmetrics"
+)
+
+var factory ClientFactory
+
+// Mutex for when modifying the stored clients
+var mutex = &sync.Mutex{}
+
+const expirationTime = time.Minute * 15
+
+// ClientFactory interface for the clientFactory object
+type ClientFactory interface {
+	GetClient(token string) (IstioClientInterface, error)
+}
+
+// clientFactory used to generate per users clients
+type clientFactory struct {
+	ClientFactory
+	baseIstioConfig *rest.Config
+	clientEntries   map[string]*clientEntry
+}
+
+// clientEntry stored the client and its last accessed timestamp
+type clientEntry struct {
+	client       IstioClientInterface
+	lastAccessed time.Time
+}
+
+// GetClientFactory returns the client factory. Creates a new one if necessary
+func GetClientFactory() (ClientFactory, error) {
+	if factory == nil {
+		// Get the normal configuration
+		config, err := ConfigClient()
+		if err != nil {
+			return nil, err
+		}
+
+		// Create a new config based on what was gathered above but don't specify the bearer token to use
+		istioConfig := rest.Config{
+			Host:            config.Host,
+			TLSClientConfig: config.TLSClientConfig,
+			QPS:             config.QPS,
+			Burst:           config.Burst,
+		}
+
+		factory, err = newClientFactory(&istioConfig, expirationTime)
+		return factory, err
+
+	}
+	return factory, nil
+}
+
+// newClientFactory allows for specifying the config and expiry duration
+// Mock friendly for testing purposes
+func newClientFactory(istioConfig *rest.Config, expiry time.Duration) (*clientFactory, error) {
+	clientEntriesMap := make(map[string]*clientEntry)
+
+	go watchClients(clientEntriesMap, expiry)
+
+	return &clientFactory{
+		baseIstioConfig: istioConfig,
+		clientEntries:   clientEntriesMap,
+	}, nil
+}
+
+// NewClient creates a new IstioClientInterface based on a users k8s token
+func (cf *clientFactory) newClient(token string) (IstioClientInterface, error) {
+	config := cf.baseIstioConfig
+
+	config.BearerToken = token
+
+	return NewClientFromConfig(config)
+}
+
+// GetClient returns a client for the specified token. Creating one if necessary.
+func (cf *clientFactory) GetClient(token string) (IstioClientInterface, error) {
+	clientEntry, err := cf.getClientEntry(token)
+	if err != nil {
+		return nil, err
+	}
+	return clientEntry.client, nil
+}
+
+// getClientEntry returns a clientEntry for the specified token. Creating one if necessary.
+func (cf *clientFactory) getClientEntry(token string) (*clientEntry, error) {
```

Comment:
I haven't tested, but under high load, we are forcing all calls to pass with the same mutex.Lock(), that sounds on paper that could be a penalty.

In the other hand, I wonder if using a default expiration could be enough, in that way, we can use a RW semaphore, and let multiple reads, only blocking when adding a new element (which can be configurable).

On this way, everytime a client is get, it's locked in the mutex.

On a expiration strategy, without marking last access time, but just the creation time, it can be more performance with the penalty to have one time re-creation of the clients.

Reply:
I've updated how we are handling the locking here.

Its now using a RWMutex. We only RLock when we fetch from the map (updating the timestamp doesn't need to be protected in a lock. We can have multiple threads concurrently updating this value).

We still need a lock when we create a new entry in the map and when we iterate over the map.

@lucasponce what do you think?

Reply:
Not sure if we can update the lastAccessed concurrently without enter in a race condition (this why the CI is failing).
And also, it could also miss the intention of your code, I guess that a check on expiration is more safer, in the worst case, a expired session but active will need to create a new client.

Maintaining the lastAccessed unprotected it can end in some undefined state (which also can lead to the same scenario) but having a race in the code.

I ask cc @simonpasquier here, as he helped us with some concurrency discussion in the past and it was good to see if something was missing in our comments.

Reply:
If we have two threads trying to concurrently set the lastAccessedTime, it doesn't really matter the order its applied. We don't need high precision here for expiration purposes.

> Maintaining the lastAccessed unprotected it can end in some undefined state (which also can lead to the same scenario) but having a race in the code.

Can it be in an undefined state? I don't believe its possible for a variable in Go to be undefined, the worst case it it would be in its default state (which is Jan 1 0000)

[Note this could actually happen in the code since we don't set the lastAccessedTime time in a mutex when we create the client, so we could have the expiration checker check the client's expiration time before its set, and since it would be year 0000 it would get marked for removal. If we want to keep the lastAccessedTIme we might want to fix this part]

There is a race condition here, but I don't believe it can actually cause a problem. Running simulations with a 10ks threads concurrently creating and deleting clients didn't have any issues.

I can remove this if we think its more safe.

@lucasponce Do we want to just remove the update to the expiration time and just expire them based on creation time? Or do you want this handled another way?

Reply:
@mwringe I think that expire them on creation time is safe.

Probably you are right and the race condition sounds "correct" for this use case, but leaving a race condition in the code creates a weird feeling.

So, if we can make it safe I'd prefer it.

We can see in a following PR if we can implement the same scenario but without having a race in the code.

Reply:
Sounds good, I will update the PR to expire based on creation time and will remove the race conditions.

Reply:
Sorry for the late reply... Basically @lucasponce is right: when the race detector reports a problem, there's 99.999% changes that it is right, false positives are extremely rare. Any race conditions can generate heisenbugs even though they seem innocuous at first.